src/logger.ts

```diff
@@ -0,0 +1,432 @@
+			this.gzip.flush(zlib.Z_SYNC_FLUSH);
+		}
+	}, ZLIB_TIMEOUT, { leading: false });
+
+	private teardown() {
+		if (this.req != null) {
+			clearTimeout(this.timeout);
+			this.req.removeAllListeners();
+			this.req.on('error', _.noop);
+			if (this.gzip != null) {
+				this.stream.unpipe(this.gzip);
+				this.gzip.end();
+			}
+			this.req = null;
+		}
+	}
+
+	private write(message: LogMessage) {
+		this.snooze();
+		if (this.req == null) {
+			this.setup();
+		}
+
+		if (this.writable) {
+			this.writable = this.stream.write(JSON.stringify(message) + '\n');
+			this.flush();
+		} else {
+			this.dropCount += 1;
+		}
+	}
+}
+
+interface LoggerConstructOptions {
+	eventTracker: EventTracker;
+}
+
+interface LoggerSetupOptions {
+	apiEndpoint: string;
+	uuid: string;
+	deviceApiKey: string;
+	offlineMode: boolean;
+	enableLogs: boolean;
+}
+
+type LogEventObject = Dictionary<any> | null;
+
+enum OutputStream {
+	Stdout,
+	Stderr,
+}
+
+export class Logger {
+	private writeLock: (key: string) => Bluebird<() => void> = Bluebird.promisify(
+		new Lock().async.writeLock,
+	);
+	private backend: LogBackend | null = null;
+	private eventTracker: EventTracker;
+	private attached: {
+		[key in OutputStream]: { [containerId: string]: boolean }
+	} = {
+		[OutputStream.Stderr]: { },
+		[OutputStream.Stdout]: { },
+	};
+
+	public constructor({ eventTracker }: LoggerConstructOptions) {
+		this.backend = null;
+		this.eventTracker = eventTracker;
+	}
+
+	public init({
+		apiEndpoint,
+		uuid,
+		deviceApiKey,
+		offlineMode,
+		enableLogs,
+	}: LoggerSetupOptions,
+	) {
+		this.backend = new ResinLogBackend(apiEndpoint, uuid, deviceApiKey);
+		this.backend.offlineMode = offlineMode;
+		this.backend.publishEnabled = enableLogs;
+	}
+
+	public enable(value: boolean = true) {
+		if (this.backend != null) {
+			this.backend.publishEnabled = value;
+		}
+	}
+
+	public logDependent(message: LogMessage, device: { uuid: string }) {
+		if (this.backend != null) {
+			message.uuid = device.uuid;
+			this.backend.log(message);
+		}
+	}
+
+	public log(message: LogMessage) {
+		if (this.backend != null) {
+			this.backend.log(message);
+		}
+	}
+
+	public logSystemMessage(
+		message: string,
+		eventObj?: LogEventObject,
+		eventName?: string,
+	) {
+		const msgObj: LogMessage = { message, isSystem: true };
+		if (eventObj != null && eventObj.error != null) {
+			msgObj.isStdErr = true;
+		}
+		this.log(msgObj);
+		this.eventTracker.track(
+			eventName != null ? eventName : message,
+			eventObj != null ? eventObj : { },
+		);
+	}
+
+	public lock(containerId: string): Bluebird.Disposer<() => void> {
+		return this.writeLock(containerId)
+			.disposer((release) => {
+				release();
+			});
+	}
+
+	public attach(
+		docker: Docker,
+		containerId: string,
+		serviceInfo: { serviceId: string, imageId: string },
+	): Bluebird<void> {
+		return Bluebird.using(this.lock(containerId), () => {
+			return this.attachStream(docker, OutputStream.Stdout, containerId, serviceInfo)
```

Comment:
Any reason this isn't a `Promise.all([attachStdout, attachStderr])`?  Do we have to attach in order?

Reply:
I'm not sure, I was keeping it the same as https://github.com/resin-io/resin-supervisor/pull/733/files#diff-0f68e03b8d0f7c87ec71275f52139ee9L213

Reply:
Ok, doesn't have to be this PR, but cc @pcarranzav do you know the answer to this?